RecoTracker/MkFit/plugins/MkFitProducer.cc

```diff
@@ -0,0 +1,148 @@
+  if (seedClean == "none") {
+    seedCleanOpt = mkfit::ConfigWrapper::SeedCleaningOpts::noCleaning;
+  } else if (seedClean == "N2") {
+    seedCleanOpt = mkfit::ConfigWrapper::SeedCleaningOpts::cleanSeedsN2;
+  } else {
+    throw cms::Exception("Configuration")
+        << "Invalida value for parameter 'seedCleaning' " << seedClean << ", allowed are none, N2";
+  }
+
+  auto backwardFitOpt =
+      backwardFitInCMSSW_ ? mkfit::ConfigWrapper::BackwardFit::noFit : mkfit::ConfigWrapper::BackwardFit::toFirstLayer;
+
+  // TODO: what to do when we have multiple instances of MkFitProducer in a job?
+  mkfit::MkBuilderWrapper::populate(isFV);
+  mkfit::ConfigWrapper::initializeForCMSSW(seedCleanOpt, backwardFitOpt, mkfitSilent_);
+}
+
+void MkFitProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add("hitsSeeds", edm::InputTag("mkFitInputConverter"));
+  desc.add<std::string>("buildingRoutine", "cloneEngine")
+      ->setComment("Valid values are: 'bestHit', 'standard', 'cloneEngine', 'fullVector'");
+  desc.add<std::string>("seedCleaning", "N2")->setComment("Valid values are: 'none', 'N2'");
+  desc.add("backwardFitInCMSSW", false)
+      ->setComment("Do backward fit (to innermost hit) in CMSSW (true) or mkFit (false)");
+  desc.addUntracked("mkfitSilent", true)->setComment("Allows to enables printouts from mkfit with 'False'");
+
+  descriptions.add("mkFitProducer", desc);
+}
+
+std::unique_ptr<mkfit::MkBuilderWrapper> MkFitProducer::beginStream(edm::StreamID iID) const {
+  return std::make_unique<mkfit::MkBuilderWrapper>();
+}
+
+namespace {
+  std::once_flag geometryFlag;
+}
+void MkFitProducer::produce(edm::StreamID iID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  const auto& hitsSeeds = iEvent.get(hitsSeedsToken_);
+  const auto& geom = iSetup.getData(geomToken_);
+
+  // First do some initialization
+  // TODO: the mechanism needs to be improved...
+  std::call_once(geometryFlag, [&geom, nlayers = hitsSeeds.nlayers()]() {
+    // TODO: eventually automatize fully
+    // For now it is easier to use purely the infrastructure from mkfit
+    /*
+      const auto barrelLayers = geom.numberOfLayers(PixelSubdetector::PixelBarrel) + geom.numberOfLayers(StripSubdetector::TIB) + geom.numberOfLayers(StripSubdetector::TOB);
+      const auto endcapLayers = geom.numberOfLayers(PixelSubdetector::PixelEndcap) + geom.numberOfLayers(StripSubdetector::TID) + geom.numberOfLayers(StripSubdetector::TEC);
+      // TODO: Number of stereo layers is hardcoded for now, so this won't work for phase2 tracker
+      const auto barrelStereo = 2 + 2;
+      const auto endcapStereo = geom.numberOfLayers(StripSubdetector::TID) + geom.numberOfLayers(StripSubdetector::TEC);
+      const auto nlayers = barrelLayers + barrelStereo + 2*(endcapLayers + endcapStereo);
+      LogDebug("MkFitProducer") << "Total number of tracker layers (stereo counted separately) " << nlayers;
```

Comment:
is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant.
The commented out part is just useful for the LogDebug call. If it's useful, there should be no harm to uncomment it

Reply:
Nah, will remove.